table/tables/mutation_checker.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tables
+
+import (
+	"fmt"
+
+	"github.com/pingcap/tidb/table"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/rowcodec"
+	"go.uber.org/zap"
+)
+
+type mutation = struct {
+	key   kv.Key
+	flags kv.KeyFlags
+	value []byte
+}
+
+type indexHelperInfo = struct {
+	indexInfo   *model.IndexInfo
+	rowColInfos []rowcodec.ColInfo
+}
+
+// CheckIndexConsistency checks whether the given set of mutations corresponding to a single row is consistent.
+// Namely, assume the database is consistent before, applying the mutations shouldn't break the consistency.
+// It aims at reducing bugs that will corrupt data, and preventing mistakes from spreading if possible.
+//
+// Assume the set of row values changes from V1 to V2. We check
+// (1) V2 - V1 = {added indices}
+// (2) V1 - V2 = {deleted indices}
+//
+// To check (1), we need
+// (a) {added indices} is a subset of {needed indices} => each index mutation is consistent with the input/row key/value
+// (b) {needed indices} is a subset of {added indices}. The check process would be exactly the same with how we generate
+// 		the mutations, thus ignored.
+func CheckIndexConsistency(sessVars *variable.SessionVars, t *TableCommon,
```

Comment:
We may need to add check and comments that this function could only be used on physical table but not partitioned table.